driver-async/src/main/com/mongodb/async/client/MongoClients.java

```diff
@@ -218,25 +224,57 @@ public static CodecRegistry getDefaultCodecRegistry() {
+        }
+        final TlsChannelStreamFactoryFactory streamFactoryFactory = new TlsChannelStreamFactoryFactory();
+        StreamFactory streamFactory = streamFactoryFactory.create(settings.getSocketSettings(), settings.getSslSettings());
+        StreamFactory heartbeatStreamFactory = streamFactoryFactory.create(settings.getHeartbeatSocketSettings(),
+                settings.getSslSettings());
+        return createMongoClient(settings, mongoDriverInformation, streamFactory, heartbeatStreamFactory,
+                new Closeable() {
+                    @Override
+                    public void close() {
+                        streamFactoryFactory.close();
+                    }
+                });
+    }
+
+    private static boolean isJava8() {
+        try {
+            Class.forName("java.time.Instant");
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    private static MongoClient createWithAsynchronousSocketChannel(final MongoClientSettings settings,
+                                                                   @Nullable final MongoDriverInformation mongoDriverInformation) {
+        StreamFactoryFactory streamFactoryFactory = new AsynchronousSocketChannelStreamFactoryFactory();
+        StreamFactory streamFactory = streamFactoryFactory.create(settings.getSocketSettings(), settings.getSslSettings());
+        StreamFactory heartbeatStreamFactory = streamFactoryFactory.create(settings.getHeartbeatSocketSettings(),
+                settings.getSslSettings());
+        return createMongoClient(settings, mongoDriverInformation, streamFactory, heartbeatStreamFactory, null);
+    }
+
+    private static StreamFactory getStreamFactory(final MongoClientSettings settings, final boolean isHeartbeat) {
         StreamFactoryFactory streamFactoryFactory = settings.getStreamFactoryFactory();
-        SocketSettings socketSettings = isHeartbeat ? settings.getHeartbeatSocketSettings() : settings.getSocketSettings();
-        if (streamFactoryFactory != null) {
-            return streamFactoryFactory.create(socketSettings, settings.getSslSettings());
-        } else if (isNio2(streamType)) {
-            return new AsynchronousSocketChannelStreamFactory(socketSettings, settings.getSslSettings());
-        } else {
-            throw new IllegalArgumentException("Unsupported stream type: " + streamType);
+        if (streamFactoryFactory == null) {
+            throw new MongoInternalException("should not happen");
         }
+        return streamFactoryFactory.create(isHeartbeat ? settings.getHeartbeatSocketSettings() : settings.getSocketSettings(),
+                settings.getSslSettings());
     }
 
     private static boolean isNetty(final String streamType) {
         return streamType.toLowerCase().equals("netty");
     }
 
-    private static boolean isNio2(final String streamType) {
+    private static boolean isNio(final String streamType) {
```

Comment:
Why rename?

Reply:
Just because TlsChannel isn't NIO 2, it's NIO 1.

Reply:
but the string is nio2?

Reply:
The string is, yes, but logically it's NIO from the perspective of what this class is actually doing.  Happy to revert if it's confusing though.